File: backend/src/api/dto/daily.py
"""Daily-game player-facing status DTOs (see services/daily_games_service.py's
DailyGamesService.get_daily_status/create_daily_game). Admin config DTOs live in api/dto/admin.py;
the leaderboard DTO lives in api/dto/leaderboard.py."""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel

from services.daily_games_service import DailyModeStatus

DailyModeStatusValue = Literal["not_played", "in_progress", "finished"]


class DailyModeStatusOut(BaseModel):
    game_type: str
    mode: str
    status: DailyModeStatusValue
    game_id: UUID | None
    score: int | None

    @classmethod
    def from_status(cls, status: DailyModeStatus) -> "DailyModeStatusOut":
        return cls(
            game_type=status.game_type,
            mode=status.mode,
            status=status.status,
            game_id=status.game_id,
            score=status.score,
        )


class DailyStatusOut(BaseModel):
    # ISO datetimes (server time) - the frontend's countdown ticks off the offset
    # between these two rather than trusting its own clock alone.
    resets_at: datetime
    server_now: datetime
    modes: list[DailyModeStatusOut]

    @classmethod
    def from_statuses(
        cls, resets_at: datetime, server_now: datetime, statuses: list[DailyModeStatus]
    ) -> "DailyStatusOut":
        return cls(
            resets_at=resets_at, server_now=server_now, modes=[DailyModeStatusOut.from_status(s) for s in statuses]
        )

File: frontend/src/api/types/daily.ts
// Roadmap #G - daily games, player-facing status. Mirrors backend/src/api/dto/daily.py (see
// services/games_service.py's GamesService.get_daily_status).

export type DailyModeStatusValue = "not_played" | "in_progress" | "finished"

export interface DailyModeStatusOut {
  game_type: string
  mode: string
  status: DailyModeStatusValue
  game_id: string | null
  score: number | null
}

export interface DailyStatusOut {
  // ISO datetimes (server time) - the countdown ticks off their offset rather than trusting the
  // client's own clock alone (see menu/DailyCountdown.tsx).
  resets_at: string
  server_now: string
  modes: DailyModeStatusOut[]
}
